src/utils/commands/CommandHandler.ts

```diff
@@ -0,0 +1,138 @@
+import { Message, PermissionString } from "discord.js";
+import Arguments from "./arguments/Arguments";
+import { join } from "path";
+import { promises as fs } from "fs";
+import Command from "./Command";
+import { MissingPermission } from "../helper/Unauthorized";
+import DiscordClient from "../client/client";
+
+interface Options {
+    dir: string;
+    prefix: string;
+}
+
+class CommandHandler {
+    private prefix: string;
+    private dir: string;
+
+    constructor(
+        private client: DiscordClient,
+        private options: Partial<Options>
+    ) {
+        this.prefix = options.prefix || ",";
+        this.dir = this.options.dir || "";
+        this.setup();
+    }
+
+    private async setup(): Promise<void> {
+        await this.loadAll();
+        //console.log(this.client.categories);
+        this.client.on("message", (msg: Message): void => {
+            if (msg.author.bot) return;
+            this.handle(msg);
+        });
+    }
+
+    private handle(message: Message): void {
+        if (!message.content.startsWith(this.prefix)) return;
+        const [commandName, ...args] = message.content
+            .slice(this.prefix.length)
+            .trim()
+            .split(/\s+/);
+
+        const command = this.client.commands.get(commandName.toLowerCase());
+
+        if (command) {
+            if (this.checkForPermissions(message, command)) {
+                const parsedArguments = new Arguments(
+                    this.client,
+                    command.options.args || [],
+                    message,
+                    [...args]
+                );
+                if (parsedArguments.error) {
+                    console.log("command error");
+                    return;
+                }
+
+                command.exec(message, parsedArguments.parsedArguments);
+            } else {
+                //TODO: implement embed to send whenever the user isn't allowed to run the command
+                //message.channel.send(`You're not the owner`);
+            }
+        }
+    }
+
+    private checkForPermissions(message: Message, command: Command): boolean {
```

Comment:
I think a better name for this function would be; "isAuthorized" or "userIsAuthorized".

Reply:
or maybe `checkForUserPermissions`

Reply:
I think that shorter version might get along better.